FR #1059: Expand Normative Requirements Guidelines with Core Authoring Rules and Attribute Requirements (#1967)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: guidelines/contributors/normative-requirements-guidelines.md

Reviewer: This PR has seemingly moved beyond the scope of just attributes.  While I fully agree with the changes reflected here, not least because it merely codifies institutional practice that we have already been encouraging/following, the name of the PR itself no longer reflects the true nature of the changes.  

Please change the PR name to be more inclusive of the true scope of the change here.

Author: I've updated the PR title from **`Establish FOCUS Attribute Specification Guidelines`** to: 
**`Expand Normative Requirements Guidelines with Core Authoring Rules and Attribute Requirements`**

Reviewer: Looks good!  Resolving.

Reviewer: @ijurica 
Please confer with @jpradocueva and determine what content from #2111 should be rolled into #1967.  I propose that pretty much everything having to do with normative requirements gets moved to #1967, and #2111 keeps a simple requirement that normative requirements are listed as bullets, and for all other details on normative requirements, to link to the normative requirements guidelines.

Reviewer: We agreed in Apr 6 Maintainers that we would look to get #1967 approved, then @jpradocueva would use that as a baseline for revising #2111 to propose further changes to normative-requirements-guidelines.md.  No action here; resolving.

Reviewer: A few more AI-gen, document-level inconsistencies to address.

### Inconsistent Formatting within the Guidelines Document

#### 1. Inconsistent Formatting of Examples (Raw Text vs. Code Blocks)
Throughout the "Column Requirements" section, examples are neatly encapsulated in standard markdown code blocks (e.g., markdown or json). However, in the new **Attribute Normative Requirements Examples** section at the very bottom of the file, the examples are pasted as raw text.

**Suggested Fix:** Wrap the four attribute examples at the bottom of the file in markdown code blocks to match the rest of the document.
```diff
 #### Null Handling
 
 > *Note: This example illustrates the baseline pattern for an Attribute with flat bullets and no qualifiers.*
 
+```markdown
 Column conforming to NullHandling attribute MUST adhere to the following requirements:
 
 * [*FOCUS dataset column*](#glossary:FOCUS-dataset-column) MUST use `null` for absent values when the *FOCUS dataset column* is defined as nullable.
 * *FOCUS dataset column* MUST NOT contain empty strings or placeholder strings (e.g., `Not Applicable`) for absent values when the *FOCUS dataset column* contains string values.
 * *FOCUS dataset column* MUST NOT contain placeholder numeric values (e.g., `0`) for absent values when the *FOCUS dataset column* contains numeric values.
+```
```

#### 2. Inconsistent Table Header Formatting
In the tables for "Logical Grouping of Dataset Requirements" and "Logical Grouping of Column Requirements," the table headers are explicitly bolded. However, the new table introduced in the **Normative Subjects in Attribute Requirements** section lacks this bolding.

**Suggested Fix:** Bold the table headers in the Attribute section so all tables in the guidelines share the same visual weight.
```diff
-| Anchor Subject Type | Requirement Subject |
+| **Anchor Subject Type** | **Requirement Subject** |
 |---|---|
 | Dataset | FOCUS dataset |
```

#### 3. Inconsistent Formatting of "Exceptions" and "Notes"
The document uses several different visual paradigms to call out exceptions and notes (e.g., bold standalone text, standard bullet points, blockquotes, and inline bolding). We should standardize on the blockquote format (`> **Exception:**` or `> **Note:**`), as it is already used effectively for the "RECOMMENDED" deprecation warning and the example preambles.

**Suggested Fix:** Standardize standalone rule exceptions to use blockquotes. (Example for Composite Requirements):
```diff
-**Exceptions:**
+> **Exception:**
 
-* **Conformance Recommendations:** When a parent bullet uses a SHOULD keyword to establish recommended conformance...
+When a parent bullet uses a SHOULD keyword to establish recommended conformance...
```

#### 4. The `* i.` List Formatting Bug (Appears in 3 Places)
In the refactor of the lists under "Logical Grouping of Dataset Requirements", "Logical Grouping of Column Requirements", and "Grouping of Attribute Requirements", nested numbered items use an asterisk followed by a roman numeral (e.g., `* i.`). In standard Markdown rendering, this creates an unordered bullet point where the text literally begins with the characters "i.", "ii.", etc. 

**Suggested Fix:** Change the `* i.` format to a properly indented numbered format in all three sections:
```diff
 1. **Presence Requirements**
-   * i. **Dataset Presence:** Defines whether, and under what conditions...
-   * ii. **Column Presence in Dataset:** Defines which columns must...
+   1. **Dataset Presence:** Defines whether, and under what conditions...
+   2. **Column Presence in Dataset:** Defines which columns must...
```

Author: Thanks for the detailed review. All four points have been addressed:

**1. Inconsistent Formatting of Examples (Raw Text vs. Code Blocks)**
Addressed. All examples across the document (dataset, column, and attribute) are now wrapped in markdown code blocks.

**2. Inconsistent Table Header Formatting**
Addressed, but differently than suggested. Instead of adding bold to the third table, I removed bold from all table headers.

**3. Inconsistent Formatting of "Exceptions" and "Notes"**
Addressed with a different approach than the suggested blockquote standardization. The document now uses a consistent inline bold pattern:
- `**Exception for {Name}:**` for named exceptions
- `**Important Note:**` for important callouts
- `**Note:**` / `**Notes:**` for informational callouts

This keeps callouts visually distinct without introducing blockquote formatting that would break the flow of the surrounding content.

**4. The `* i.` List Formatting Bug**
Addressed. Using standard nested ordered lists (with the 1. prefix) to allow for proper automatic numbering

---

@shawnalpay, could you please take a look and let me know if these thread can be resolved?

Reviewer: I'm not onboard with how you've done issue 3, because it's not spelled out this way in the editorial guidelines or `AGENTS.md`.  But I'm certainly not prepared to block on that.  I've added [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/2111/files#r3053171869) to #2111 for us to discuss further.

The other issues look good!  Resolving.

* **FOCUS Column**, whereby use of:  
  * `FOCUS column` keyword represents any FOCUS column  
  * `FOCUS column` keyword with a qualifier represents a qualified subset of FOCUS columns (e.g., `FOCUS column containing numeric values`)  
  * A single FOCUS column explicitly identified by `<FOCUS Column ID>` (e.g., `BilledCost`)

* **Custom Column**, whereby use of:  
  * `Custom column` keyword represents any custom column  
  * `Custom column` keyword with a qualifier represents a qualified subset of custom columns (e.g., `Custom column containing numeric values`)

* **Structural sub-elements within Columns** (objects, keys, key values):  
  *Note: MUST NOT use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  
  * `Key in Object in [FOCUS|Custom] column containing JsonObjectFormat values`  
  * `Key value in Object in [FOCUS|Custom] column containing key-value pairs`
  * `Object in FOCUS dataset column`  
  * `Object in array in FOCUS dataset column`  
  * `Key in Object in FOCUS dataset column`  

>   * `Key value in Object in FOCUS dataset column`  

Reviewer: This section needs to be revisited, given that we have a dedicated Object with a display name and an ID.

The following columns now adhere to the agreed format in `working_draft`.  This guidelines need to align with this format.

* [AllocatedMethodDetails](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/allocatedmethoddetails.md)
* [ContractApplied](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/cost_and_usage/columns/contractapplied.md)
* [ContractCommitmentApplicability](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/datasets/contract_commitment/columns/contractcommitmentapplicability.md)

When I give an LLM these guidelines and these columns and say, "Are there any changes to the format we need to make to the columns, per the guidelines?", the LLM currently wants to make heavy revisions.  We cannot resolve this comment until we revise the guidelines to significantly reduce the gap between guidelines and execution.

Author: ### Additional Guidelines for Columns in JSON Format

I've updated the Additional Guidelines for Columns in JSON Format section. The main changes are:

- Split the section into dedicated **Key-Value Format Columns** and **JSON Object Format Columns** subsections, as the two formats have distinct conventions for structuring requirements (single Requirements section vs. Column Requirements + Object Requirements, key/value terminology vs. dot-notation, etc.).
- Added Schema Requirements patterns for JSON Object Format columns, covering both object conformance in Column Requirements and JSON Schema reference in Object Requirements.
- Added Object Properties guidance covering dot-notation, singular form, and aggregate expressions.
- Removed JTD (JSON Type Definition) references and structural pattern examples, as current columns do not use JTD and the JSON examples belong in the columns themselves, not in the guidelines.

@shawnalpay, @kk09v  could you please take a look?

---

#### The aggregate requirement in `AllocatedMethodDetails`

The following requirement doesn't follow the Allowed Subjects exception for aggregate expressions used elsewhere in the spec (e.g., `The sum of {ColumnId} in a given billing period...`).

> * Values for all AllocatedMethodDetailsObject.Elements[*].AllocatedRatio properties across all allocated charges related to a single origin charge MUST sum up to 1 (100%).

I'd suggest rephrasing it to: 

> * The sum of AllocatedMethodDetailsObject.Elements[*].AllocatedRatio across all allocated charges related to a single origin charge MUST equal 1 (100%).

@shawnalpay @kk09v @qquigley thoughts? if yes, we can add this to the #1878 queue.

Reviewer: @ijurica 
The guidelines look great now, thank you!  

I'm onboard with changing it to `The sum of...` per the exception you added to the Allowed Subjects.

Leaving this open for you to collect more feedback during TF-2 today, but as far as I'm concerned, you can resolve.

Reviewer 2: Reviewed that section and it looks good to me. One thing I wanted to clarify is the term "key value" refers to the value associated with a specific key, correct? Currently lines 146, 150, 152; lines 135 and 139 in the outdated comment.
I.e. if we've said that the key needs to be CerealType, the structural sub-element would refer to the "CerealType value"

Edit: After reading through the rest of the PR, if my understanding of both pieces is correct (big if), should it be "<key> value" here?

Reviewer 2: I also agree with the rephrasing in Allocated Method Details

Author: previously mentioned suggestion added to the #1878 queue (see [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1878#issuecomment-4214742554))

---

`Key value` does refer to the value associated with a specific key. The initial list of subject forms (e.g., `Key in Object in FOCUS dataset column`, `Key value in Object in FOCUS dataset column`) covers the generic forms used in Attribute requirements, but was missing the column-specific patterns 😞 

Thanks for flagging! 

I've expanded the examples with column-specific subject forms:

> * `<term> key` / `<term> value` for column-specific key-value terminology (e.g., `Tag keys`, `Property value`)
> * `<ObjectId>` for JSON Object-level requirements (e.g., `ContractAppliedObject`)
> * `<ObjectId>.<PropertyPath>` for JSON Object property-level requirements (e.g., `ContractCommitmentApplicabilityObject.Applicability.Cost`)
> * `<ObjectId>.<PropertyPath>[*].<PropertyPath>` for properties within arrays (e.g., `ContractAppliedObject.Elements[*].ContractId`, `ContractCommitmentApplicabilityObject.Inclusions[*].Dimension`)

@kk09v please check and let me know if can resolve this thread.

Reviewer: This looks good to me, and given that we 1) have Member approval pending later today, and 2) @kk09v has now approved the PR, I'm going to resolve the thread.  If something still looks off, @kk09v can reopen.  Thanks @ijurica!

  *Example: `ListCost MUST equal the product of ListUnitPrice and PricingQuantity when ListUnitPrice is not null and PricingQuantity is not null.` — this requirement is defined on ListCost. ListUnitPrice and PricingQuantity MAY reference it in their introductory description but MUST NOT restate it as a normative requirement.*

* If a requirement spans multiple datasets, it MUST be defined on the column in the dataset that is the primary owner of the validation. Other datasets involved MUST NOT restate it as a normative requirement but MAY reference it in their introductory description.

  *Example: A cross-dataset sum validation comparing BilledCost aggregated by InvoiceId and InvoiceIssuerName between InvoiceDetail and CostAndUsage is defined on InvoiceDetail.BilledCost, as InvoiceDetail is the primary owner of invoice-level validation. CostAndUsage MAY reference it in its introductory description but MUST NOT restate it as a normative requirement.*

## Dataset Requirements

> ### Logical Grouping of Dataset Requirements

Reviewer: ### Markdown List Formatting Bug

In the refactor of these lists, you changed the nested numbered items to use an asterisk followed by a roman numeral (e.g., `* i.`). 

In standard Markdown rendering, this does not create a nested ordered list. It creates a standard unordered bullet point where the text literally begins with the characters "i.", "ii.", etc. If the goal is a nested ordered list, you should use standard indentation with numbers (or letters, depending on the parser).

**Suggested Fix:**
Change the `* i.` format to a properly indented ordered format:
```diff
   1. **Presence Requirements**
-     * i. **Dataset Presence:** Defines whether, and under what conditions...
-     * ii. **Column Presence in Dataset:** Defines which columns must...
+     1. **Dataset Presence:** Defines whether, and under what conditions...
+     2. **Column Presence in Dataset:** Defines which columns must...
```
*(Apply this to both the Dataset and Column requirement grouping lists)*

Author: Updated all three grouping sections. using standard nested ordered lists (with the 1. prefix) to allow for proper automatic numbering

Reviewer: Looks good, thank you!

|---|---|---|---|
| Presence | Dataset Presence | Always | {DatasetId} MUST be present when {Condition}. |
| Presence | Column Presence in Dataset | Always | {DatasetId} MUST include {ColumnId}. |
| Attribute Conformance | Dataset Attribute Conformance | Always | {DatasetId} MUST conform to DatasetCompleteness requirements. |
| Attribute Conformance | FOCUS Column Attribute Conformance | Always | {DatasetId} *FOCUS columns* MUST conform to NullHandling requirements. |
| Attribute Conformance | Custom Column Attribute Conformance | Always | {DatasetId} *custom columns* MUST conform to CustomColumnHandling requirements. |
| Other | Other | When applicable | InvoiceDetail MUST represent all invoice line items with a non-zero BilledCost on any invoice associated with a BillingAccountId. |

### Ordering of Dataset Requirements Within Groups

To further enhance readability, individual requirements within each group SHOULD be ordered as follows:

* **MUST** – an absolute requirement
* **MUST NOT** – a prohibition
* **SHOULD** – recommended but not mandatory
* **SHOULD NOT** – discouraged but not strictly prohibited
* **MAY** – optional

> * **MAY NOT** – optional prohibition / permitted not to

Reviewer: ### Exception for Column Presence Ordering

The column presence requirements (e.g., `MUST include...`, `SHOULD include...`) are sorted alphabetically by Column ID. This makes common sense. 

However, under the current, strict rules in `normative-requirements-guidelines.md`, this technically violates the BCP 14 keyword ordering rule, because a `SHOULD include` (like `AllocatedMethodDetails`) ends up sitting above a `MUST include` (like `AllocatedResourceId`).

To facilitate this intuitive alphabetical sorting without breaking the parsing rules, we should add an explicit exception to the guidelines for the **Column Presence in Dataset** group.

**Suggested Fix:**
```suggestion
* **MAY NOT** – optional prohibition / permitted not to

* **Exception for Column Presence:** Requirements within the **Column Presence in Dataset** group MUST be ordered alphabetically by the referenced Column ID, taking precedence over the BCP 14 keyword ordering.
```

| Dataset | Other Dataset Requirements | When applicable | CostAndUsage MUST have its data generator-calculated split cost allocation method documented and accessible to practitioners. |
| FOCUS columns | FOCUS Column Attribute Conformance | Always | {DatasetId} *FOCUS columns* MUST conform to NullHandling requirements. |
| FOCUS columns | Other FOCUS Column Requirements | When applicable | |
| Custom columns | Custom Column Attribute Conformance | Always | {DatasetId} *custom columns* MUST conform to CustomColumnHandling requirements. |
| Custom columns | Other Custom Column Requirements | When applicable | |
| Documentation | Documentation | When applicable | {DatasetId} documentation MUST specify how records correspond to invoice line items. |
| Other | Other | When applicable | |

### Ordering of Dataset Requirements Within Groups

To further enhance readability, individual requirements within each group SHOULD be ordered as follows:

* **MUST** – an absolute requirement
* **MUST NOT** – a prohibition
* **SHOULD** – recommended but not mandatory
* **SHOULD NOT** – discouraged but not strictly prohibited
* **MAY** – optional

> * **MAY NOT** – optional prohibition / permitted not to

Reviewer: **Inclusion of `MAY NOT` is a violation of BCP 14 (RFC 2119).** RFC 2119 explicitly defines `MUST`, `MUST NOT`, `REQUIRED`, `SHALL`, `SHALL NOT`, `SHOULD`, `SHOULD NOT`, `RECOMMENDED`, `MAY`, and `OPTIONAL`. 

It intentionally omits `MAY NOT` because it is highly ambiguous in the English language. Depending on the context, "may not" can mean "is strictly prohibited from" (e.g., "You may not enter") or "is permitted to not" (e.g., "You may or may not do this"). Because of this ambiguity, IETF standards strictly avoid it.

**Suggested Fix:**
Please remove `MAY NOT` from the allowed keywords lists in the guidelines entirely. If the spec needs to say "permitted not to," it should be rephrased using valid BCP 14 terms (like stating an action is `MAY`, implying the inverse is also allowed, or avoiding the phrasing altogether).

```suggestion
* **MAY** – optional
```

> **Important clarification**

By glossary definition, the following concepts are used:

* **FOCUS Dataset** — the primary dataset concept defined by the FOCUS specification.
* **Dataset Instance** — represents a concrete instantiation of a **FOCUS Dataset Instance**.
* **Dataset Artifact** — represents a physical or delivered form of a **FOCUS Dataset Instance Artifact**.

However, by design decision, the specification adopts the following normative conventions:

* **FOCUS Dataset is used as the canonical normative subject** for dataset-level requirements.
* Normative requirements are intentionally written against **FOCUS Dataset**, even when the constraint applies to:
  * a dataset specification,
  * a dataset instance, or
  * a dataset instance artifact.
* The intended level of application (specification vs. instance vs. artifact) is inferred from context rather than encoded in the normative subject.

This choice is intentional and overrides interpretations based solely on abstraction level.

> ## 1. Core Normative Authoring Rules

Reviewer: ### Hardcoded Section Numbers

Currently, the guidelines use manually typed, deep hierarchical numbers (e.g., `3.7.1.1`). In a "docs-as-code" workflow, this creates a major maintenance bottleneck. 

If someone adds a new rule between `2.1` and `2.2`, they (and the reviewers) now have to manually renumber `2.3`, `2.4`, `2.5`, and potentially dozens of nested sub-bullets. This leads to:
* **Noisy Diffs:** Pull requests will be inflated with dozens of irrelevant lines just for number increments.
* **Merge Conflicts:** If two people are working on the file simultaneously, their numbering will collide.
* **Brittle Links:** Cross-references like *"See section 3.4.2"* will break the moment a section is inserted above them.

**The Fix: Semantic Headers & Anchor Links**
1. **Strip the numbers:** Remove the hardcoded numbers and rely entirely on standard Markdown header hierarchy (`##`, `###`, `####`).
2. **Link by name, not number:** Update cross-references to use semantic anchor links instead of numbers.
   * ✘ **Instead of:** "See section 3.4.2 for JSON requirements."
   * ✔ **Use:** "See the [JSON Schema Requirements](#json-schema) section."

Author: Addressed. The file now uses semantic Markdown headers with anchor links throughout, with no hardcoded section numbers.

@shawnalpay, could you please take a look and let me know if this can be resolved? thanks!

Reviewer: Looks great, thank you!

* contains a hidden constraint expressed as a definition (e.g., "ColumnA MUST be Z, where Z is defined as Y"),
* applies constraints to multiple subjects, even with a single BCP 14 keyword (e.g., "ColumnA and ColumnB MUST be X").

### 1.7. Composite Requirements

Composite (parent + nested) requirements MAY be used when strictly controlled.

Composite requirements MUST adhere to the following requirements:

* Nested bullets MUST share the same condition if defined by the parent bullet.
* Nested bullets SHOULD NOT introduce a different subject.
* When nested bullets introduce different subjects, all subjects SHOULD be of the same subject type (e.g., FOCUS columns).

Composite requirements SHOULD be used when grouping improves readability and:

* multiple requirements share the same subject, or
* multiple requirements share the same business context (e.g., all requirements related to a specific business scenario or feature)

Flat parallel bullets SHOULD be preferred when ordering keyword is sufficient for clarity and readability.

> 

Author: ```suggestion
Composite (parent + nested) requirements MAY be used to group related constraints under a shared condition, context, or subject.

Composite requirements MUST adhere to the following requirements:

* **Nuanced Obligation:** When a parent bullet uses a BCP 14 keyword (e.g., MUST), it establishes a mandatory requirement to evaluate the nested constraints. Each nested bullet then defines the specific nuance of that obligation for its respective subject or condition using its own BCP 14 keyword.
* **Shared Conditionality:** Nested bullets MUST share the same condition if defined by the parent bullet.
* **Context and Subject Consistency:** Nested bullets SHOULD maintain a consistent business context. While nested bullets SHOULD NOT introduce a different subject type, they MAY reference different subjects (e.g., a FOCUS dataset and its custom columns) provided they all relate to the same primary business context defined by the parent bullet.

**Exceptions:**

* **Conformance Recommendations:** When a parent bullet uses a SHOULD keyword to establish recommended conformance to a set of requirements (e.g., in `CustomColumnHandling` or when a column declares conformance to an attribute like `UnitFormat`), the weakest keyword in the hierarchy applies to the overall conformance.

Composite requirements SHOULD be used when grouping improves readability and:

* Multiple requirements share the same Business Context.
* Multiple requirements share the same subject.

Flat parallel bullets SHOULD be preferred when ordering keywords alone is sufficient for clarity and readability.

```

  * `Object in array in FOCUS dataset column`  
  * `Key in Object in FOCUS dataset column`  
  * `Key value in Object in FOCUS dataset column`  
  * `Key in FOCUS dataset column`  
  * `Key value in FOCUS dataset column`

The subject SHOULD be explicit and unambiguous.

#### 1.3.2. Disallowed Subjects

The following MUST NOT be used as normative subjects:

* Actors (e.g. data generator, service provider, consumer)
* Processes or mechanisms (e.g. Delivery Handling, Correction Handling, etc.)

### 1.4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be expressed as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 1.5. Use of BCP 14 Keywords

* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT). See [BCP14](https://tools.ietf.org/html/bcp14) [[RFC2119](https://tools.ietf.org/html/rfc2119)][[RFC8174](https://tools.ietf.org/html/rfc8174)].

> * A bullet containing more than one normative keyword MUST be split.

Author: ```suggestion
* A bullet containing more than one normative keyword MUST be split.
* **Exception for Composite Requirements:** While each individual bullet (parent or nested) MUST contain only one BCP 14 keyword, a Composite Requirement as a whole MAY contain multiple keywords to express nuanced obligations. In such cases, the logical strength of the requirement is governed by the hierarchy defined in section 1.7. Composite Requirements.
```

**Exceptions:**

* **Conformance Recommendations:** When a parent bullet uses a SHOULD keyword to establish recommended conformance to a set of requirements (e.g., in `CustomColumnHandling` or when a column declares conformance to an attribute like `UnitFormat`), the weakest keyword in the hierarchy applies to the overall conformance.

Composite requirements SHOULD be used when grouping improves readability and:

* Multiple requirements share the same Business Context.
* Multiple requirements share the same subject.

Flat parallel bullets SHOULD be preferred when ordering keywords alone is sufficient for clarity and readability.

### 1.8. Definitions vs. Normative Requirements

* Definitions, explanations, rationale, and examples MUST NOT be expressed as normative requirements.
* Definitions SHOULD be written as plain declarative statements without BCP 14 keywords.

> * Normative bullets SHOULD be reduced to the enforceable constraint only.

Author: ```suggestion
### 1.8. Contextual Information (e.g., Definitions, Examples) and Normative Authority (Requirements)

While normative requirements MUST focus on **enforceable constraints** and **verifiable states**, definitions, informative clauses, and examples MAY be included within a requirement where necessary to provide essential context and ensure unambiguous interpretation.

#### Separation of Concerns:
* **Definitions:** If a definition is complex or applies to multiple requirements, it SHOULD be placed in the [Glossary](#glossary) or the preamble section and referenced as a link within the requirement.
* **Complex Logic:** If an informative or normative clause is complex or applies to multiple requirements, it SHOULD be placed in the **Implementation Context** section to maintain the clarity of the core requirement.
* **Normative Authority:** To ensure consistency, BCP 14 keywords MUST ONLY be used within the **Requirements** section. The content in the **Glossary**, preamble, or **Implementation Context** MUST NOT contain BCP 14 keywords.

#### Non-Normative Examples:
* **Incorporation:** Examples incorporated in requirements MUST be clearly identified using "e.g." and placed within parentheses `(e.g., ...)` to distinguish them from the normative constraint.
```

# Normative Requirements Guidelines

This section defines guidelines for authoring normative requirements in the FOCUS specification. These guidelines define **how** to write normative requirements to ensure clarity, consistency, and testability. It does not define the requirements themselves (the "what") but concentrates on their **structure, subjects, and verifiability**.

The guidelines cover normative requirements applicable to:

* **FOCUS Datasets** — the primary containers of structured data as defined in FOCUS.
* **Columns** — individual columns within datasets (may contain nested objects and object properties, which can have additional normative rules)

> * **Attributes** — schema-level rules that datasets, columns, or object properties must conform to.

Reviewer 3: ```suggestion
* **FOCUS Column Attributes** — schema-level rules that FOCUS datasets, FOCUS columns, or object properties must conform to.
```

Author: I’ve slightly revised this paragraph. 

Sharing the current version below for reference:

> The guidelines cover authoring of normative requirements for the following entities:
> 
> * **FOCUS datasets** — the primary containers of structured data as defined in FOCUS.
> * **FOCUS columns** — individual columns within FOCUS datasets, defined by FOCUS (may contain nested objects and object properties, which can have additional normative rules).
> * **Custom columns** — individual columns within FOCUS datasets, not defined by FOCUS.
> * **FOCUS attributes** — reusable sets of normative constraints that datasets, columns, or column sub-elements (such as objects and object properties) conform to; guidelines cover how to author requirements within Attribute sections.

@thecloudman, please let me know if this works for you, and if I can close this thread.

---

BTW, alongside the existing glossary definition for FOCUS dataset, we introduced the following glossary terms as well:

> <a name="glossary:FOCUS-column"><b>FOCUS Column</b></a>
> 
> A column defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). See the Columns section of each dataset for definitions.
> 
> <a name="glossary:custom-column"><b>Custom Column</b></a>
> 
> A column not defined by FOCUS and included in a [*FOCUS dataset*](#glossary:FOCUS-dataset). Custom columns are prefixed with `x_` and provide additional context from [*native datasets*](#glossary:native-dataset) beyond what is captured in FOCUS columns. See [Dataset Completeness](#attributes.datasetcompleteness) for inclusion requirements.

Reviewer: I believe we have resolved this.  @thecloudman, please confirm -- otherwise we will resolve by EOD today.

Reviewer 3: This is good, @ijurica i will resolve the comment

## Core Normative Authoring Rules

### 1. Normative Requirement Structure

The recommended pattern for a normative requirement is:

``` markdown
<Subject (+qualifier)> + <BCP 14 Keyword> + <Verifiable State Descriptor> + <Object (+qualifier)> [+ Conditions]
```

* Each normative requirement MUST:
  * identify exactly one **normative subject** to which the requirement applies
  * contain exactly one **BCP 14 keyword** (MUST, SHOULD, MAY, MUST NOT, etc.), indicating the obligation level
  * express exactly one **verifiable constraint**
* Each normative requirement SHOULD describe a **verifiable state** of the object rather than behavior

### 2. Structural Anchor Requirement

Each Requirements section for a schema-level construct MUST begin with a single **structural anchor requirement**.

The structural anchor requirement:

* introduces the scope of the subsequent normative requirements,
* MUST appear as the first normative statement in the section,
* exists to support automated parsing and validation, and
* is non-verifiable and does not introduce an enforceable constraint.

The canonical form of a structural anchor requirement is:

> > `<Entity> MUST adhere to the following requirements:`

Reviewer: Disagree strongly with this requirement.  The attribute does not adhere to any requirements.  It _applies_ requirements to entities.

In PR #1981, we can see the awkward manifestation of this requirement:

> ColumnHandling MUST adhere to the following requirements:
> 
> * FOCUS column MUST adhere to the following requirements:
>   * FOCUS column MUST use Pascal case in the Column ID.

I suspect that this was done to facilitate ease of requirements model implementation.  (If you could point me to the rationale, that would be helpful.)  But setting that aside, and reading this definition in a vacuum, this reads both awkwardly and inaccurately to me.

If we truly must retain the composite, then how about:

> * The FOCUS column to which ColumnHandling applies MUST adhere to the following requirements:

Reviewer 3: ColumnHandling MUST adhere to the following requirements:

Columns in a FOCUS dataset MUST adhere to the following requirements:
FOCUS Columns MUST use Pascal case in the Column ID.

Author: Here is an example with multiple Subjects.

> JsonObjectFormat MUST adhere to the following requirements:
> 
> * FOCUS column containing JsonObjectFormat values MUST adhere to the following requirements:
>   * FOCUS column containing JsonObjectFormat values MUST contain a serialized JSON string, consistent with the [ECMA 404](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf) definition of an object.
>   * FOCUS column containing JsonObjectFormat values MUST conform to all requirements of the corresponding column definition, which may specify or restrict the shape or contents of the object.
> * Object in FOCUS column containing JsonObjectFormat values SHOULD NOT exceed 3 levels of nesting.
> * Key in Object in FOCUS column containing JsonObjectFormat values MUST be unique.
> * Key value in Object in FOCUS column containing JsonObjectFormat values MUST be of type number, string, boolean (`true` or `false`), array, object, or `null`.
> * Object in array in FOCUS column containing JsonObjectFormat values MUST adhere to the following requirements:
>   * Object in array in FOCUS column containing JsonObjectFormat values MUST be of a consistent type.
>   * Object in array in FOCUS column containing JsonObjectFormat values MUST NOT be repeated.
>   * Object in array in FOCUS column containing JsonObjectFormat values MUST NOT be null.
> * Custom column containing JSON object MUST have its object schema documented by the data generator and accessible to practitioners.

I'm open to suggestions and I can easily:
- rephrase the anchor requirement
- rename the subjects
- transform qualifier to conditions

Reviewer 4: **Agree with Shawn's framing here.** Per the guidelines themselves, "normative requirements defined in an Attribute section are enforced on the conforming schema-level entity, not on the Attribute itself" — and attributes are explicitly not normative subjects. 

So, "ColumnHandling MUST adhere to..." misstates what's actually happening. 

Shawn's proposed wording — 

"The FOCUS column to which [Attribute] applies MUST adhere to..." — correctly identifies the schema-level entity as the subject. 

Graham's simplification ("Columns in a FOCUS dataset MUST adhere to...") is **cleaner** but loses the explicit connection to the attribute doing the applying, which matters for readability in context. 

I'd go with Shawn's version.

Author: Parallel discussion on the same topic is ongoing in PR #1981. The new canonical form for Anchor Requirements for Attributes has been defined as:

``` markdown
[Dataset|Column] conforming to <Attribute ID> attribute MUST adhere to the following requirements:
```

In the other PR, the current proposal has been accepted — see [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1981/files/02b8dd391c5e7731f1c1ab1d97d4c6b67dde203b#r2937213758) for context.

Therefore, I’m closing this thread.

Note: Further adjustments are possible — some aspects may look fine on paper, but we will see how they work in practice in PR #1981.

* 🟥 FOCUS normative rule set (not a normative subject)

**Relationships:**

* `|| -- has -- |{` : one parent to one-or-more enumerated structural members
* `|| -- contains -- o{` : one parent to zero-or-more child entities (array of objects)
* `}| .. conforms-to .. ||` : many children to one parent conformance relationship

## Note on FOCUS Dataset, Dataset Instance, and Glossary Alignment

> **Important clarification**

By glossary definition, the following concepts are used:

* **FOCUS Dataset** — the primary dataset concept defined by the FOCUS specification.
* **Dataset Instance** — represents a concrete instantiation of a **FOCUS Dataset Instance**.
* **Dataset Artifact** — represents a physical or delivered form of a **FOCUS Dataset Instance Artifact**.

However, by design decision, the specification adopts the following normative conventions:

* **FOCUS Dataset is used as the canonical normative subject** for dataset-level requirements.
* Normative requirements are intentionally written against **FOCUS Dataset**, even when the constraint applies to:
  * a dataset specification,
  * a dataset instance, or
  * a dataset instance artifact.

> * The intended level of application (specification vs. instance vs. artifact) is inferred from context rather than encoded in the normative subject.

Author: To support this approach and avoid ambiguity:

* The **FOCUS Dataset** glossary definition should be refined so that it is no longer interpreted as purely abstract, but as a normative anchor that may represent the dataset specification and its specific implementations.
 * Glossary entries for **Dataset Instance** and **Dataset Artifact** SHOULD explicitly reflect that they represent **FOCUS Dataset Instance** and **FOCUS Dataset Instance Artifact**, respectively (important with the planned **Native Dataset** entry in version 1.4).

---

Additional suggestions for glossary entries:

* **FOCUS column**: Column defined by FOCUS and included in a FOCUS dataset
* **Custom column**: Column not defined by FOCUS, prefixed with `x_` and included in a FOCUS dataset
* **Native column**: Column included in a Native dataset

Reviewer: @ijurica will leave a note on #1878 to potentially track doing this during consistency review, then resolve this thread.

Reviewer: @ijurica 
Have you yet left a note on #1878 for this?  Would like to resolve this today.

Author: added a [note ](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1878#issuecomment-4165377184)in #1878

resolving

This choice is intentional and overrides interpretations based solely on abstraction level.

## Core Normative Authoring Rules

### 1. Normative Requirement Structure

The recommended pattern for a normative requirement is:

``` markdown
<Subject (+qualifier)> + <BCP 14 Keyword> + <Verifiable State Descriptor> + <Object (+qualifier)> [+ Conditions]

```

> * Each normative requirement MUST:

Reviewer 5: 🤖 [AI][Claude] Same question: Using "MUST" to describe how to write normative requirements creates "meta-normative" rules. Is this intentional? If so, it may be worth adding a note at the top of this document explaining that it uses normative language to enforce authoring consistency.

Author: See [this response](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/files#r2801492090)

Author: For consistency, we are already using BCP-14 keywords across other guidelines as well.

We could revisit this decision, and there is likely room for improvement and better consistency across the board. I suggest we do so in a separate discussion or PR, as this would affect all guidelines - not just this PR.

With that in mind, I’m resolving this thread.

  * identify exactly one **normative subject** to which the requirement applies
  * contain exactly one **BCP 14 keyword** (MUST, SHOULD, MAY, MUST NOT, etc.), indicating the obligation level
  * express exactly one **verifiable constraint**

> * Each normative requirement SHOULD describe a **verifiable state** of the object rather than behavior

Reviewer 5: 🤖 [AI][Claude] Question: This guideline uses BCP-14 keywords (SHOULD) to describe how requirements should be written. Since guidelines are non-normative documents, is the use of uppercase SHOULD intentional here? Consider using lowercase "should" to distinguish from specification requirements, or confirm this "meta-normative" usage is intentional.

Author: The use of BCP-14 keywords here was intentional. 

Regarding specification requirements, those are located in the `../specification/` folder, whereas the guidelines reside in `/guidelines/`, so I don't anticipate any confusion there.

That said, the term "guidelines" implies non-normative content, so the uppercase SHOULD/MUST/etc. may seem a bit out of place. I still suggest that we use BCP-14 keywords here.

Author: For consistency, we are already using BCP-14 keywords across other guidelines as well. With that in mind, I'm resolving this thread. 

If you feel we should revisit this decision, I suggest we do so in a separate discussion or PR, as this would affect all guidelines - not just this PR.

  * `Object in Columns containing JsonObjectFormat values`  
  * `Key in Object in FOCUS/Custom column containing JsonObjectFormat values`  
  * `Key value in Object in FOCUS/Custom column containing key-value pairs`

The subject SHOULD be explicit and unambiguous.

**Exception:** A structural anchor requirement MAY use the ID of the enclosing schema construct as its subject, even if that construct is not otherwise an allowed normative subject (e.g., `<FOCUS Attribute  ID>`). This exception applies only to structural anchor requirements.

#### 3.2 Disallowed Subjects

The following MUST NOT be used as normative subjects:

* Actors (e.g. data generator, service provider, consumer)
* Processes or mechanisms (e.g. Delivery Handling, Correction Handling, etc.)

### 4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be reformulated as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

> ### 5. Use of BCP 14 Keywords

Reviewer 6: *Revised* Pulled links from later parts in the document and moved them to the end of this section!

Author: addressed. resolving this thread

> * Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT).

Reviewer 6: ```suggestion
* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT). See [BCP14](https://tools.ietf.org/html/bcp14) [RFC2119](https://tools.ietf.org/html/rfc2119)][RFC8174](https://tools.ietf.org/html/rfc8174)].
```

Author: thanks!

* **Structural sub-elements within Columns** (objects, keys, key values):  
  *Note: MUST NOT use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  
  * `Object in Columns containing JsonObjectFormat values`  
  * `Key in Object in FOCUS/Custom column containing JsonObjectFormat values`  
  * `Key value in Object in FOCUS/Custom column containing key-value pairs`

The subject SHOULD be explicit and unambiguous.

#### 1.3.2. Disallowed Subjects

The following MUST NOT be used as normative subjects:

* Actors (e.g. data generator, service provider, consumer)
* Processes or mechanisms (e.g. Delivery Handling, Correction Handling, etc.)

### 1.4. State, Not Behavior

Normative requirements MUST describe a **verifiable state**, not an operational process or behavior.

Specifically:

* Process-oriented verbs such as *ensure*, *handle*, *support*, or *provide* MUST NOT be used.
* If a requirement refers to actor behavior, it MUST be expressed as:
  * a constraint on the resulting dataset state, or
  * a constraint on a schema-defined artifact.

### 1.5. Use of BCP 14 Keywords

> * Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT). See [BCP14](https://tools.ietf.org/html/bcp14) [RFC2119](https://tools.ietf.org/html/rfc2119)][RFC8174](https://tools.ietf.org/html/rfc8174)].

Reviewer 7: ```suggestion
* Each normative bullet MUST contain exactly one BCP 14 keyword (MUST, SHOULD, MAY, MUST NOT, SHOULD NOT). See [BCP14](https://tools.ietf.org/html/bcp14) [[RFC2119](https://tools.ietf.org/html/rfc2119)][[RFC8174](https://tools.ietf.org/html/rfc8174)].
```

Broken markdown link. Missing outer brackets on RFC references. Correct pattern already appears on lines 236, 359, and 845.

* A bullet containing more than one normative keyword MUST be split.

### 1.6. Splitting Requirements

A requirement MUST be split into multiple bullets if it:

* combines multiple obligations,
* combines a rule and an exception,
* mixes a definition with a constraint,
* applies different constraints to different subjects.

### 1.7. Composite Requirements

> Composite (parent + nested) requirements MAY be used when strictly controlled and MUST adhere to the following requirements:

Reviewer 3: This sentence contains two BCP 14 keywords (MAY and MUST), which conflicts with rule 1.5 ("A bullet containing more than one normative keyword MUST be split"). Suggest splitting into two sentences:

```suggestion
Composite (parent + nested) requirements MAY be used when strictly controlled.

Composite requirements MUST adhere to the following requirements:
```

Author: this suggestion was applied in [this commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/commits/36bc1ddfb76812d31ea1e6398db7368b8399e85f) but the thread wasn't automatically resolved.

* **FOCUS Dataset**, whereby use of:  
  * `FOCUS dataset` keyword represents any FOCUS dataset  
  * `FOCUS dataset` keyword with a qualifier represents a qualified subset of FOCUS datasets  
  * A single FOCUS dataset explicitly identified by `<FOCUS Dataset ID>` (e.g., `CostAndUsage`)

* **FOCUS Column**, whereby use of:  
  * `FOCUS column` keyword represents any FOCUS column  
  * `FOCUS column` keyword with a qualifier represents a qualified subset of FOCUS columns (e.g., `FOCUS column containing numeric values`)  
  * A single FOCUS column explicitly identified by `<FOCUS Column ID>` (e.g., `BilledCost`)

* **Custom Column**, whereby use of:  
  * `Custom column` keyword represents any custom column  
  * `Custom column` keyword with a qualifier represents a qualified subset of custom columns (e.g., `Custom column containing numeric values`)

* **Structural sub-elements within Columns** (objects, keys, key values):  
  *Note: MUST NOT use `object`, `key`, or `value` keywords alone. Always reference them in context, e.g.:*  
  * `Object in Columns containing JsonObjectFormat values`  
  * `Key in Object in FOCUS/Custom column containing JsonObjectFormat values`  

>   * `Key value in Object in FOCUS/Custom column containing key-value pairs`

Reviewer: @ijurica 
The current version of AllocatedMethodDetails, ContractApplied, and ContractCommitmentApplicability are in violation of this rule.  I argue that the way we have it is better than having it be spelled out in English the way the guideline here reads.  Are you open to changing it?

Reviewer: Will track this in consistency review.  Resolving.

  *Example: `ListCost MUST equal the product of ListUnitPrice and PricingQuantity when ListUnitPrice is not null and PricingQuantity is not null.` — this requirement is defined on ListCost. ListUnitPrice and PricingQuantity MAY reference it in their introductory description but MUST NOT restate it as a normative requirement.*

* If a requirement spans multiple datasets, it MUST be defined on the column in the dataset that is the primary owner of the validation. Other datasets involved MUST NOT restate it as a normative requirement but MAY reference it in their introductory description.

  *Example: A cross-dataset sum validation comparing BilledCost aggregated by InvoiceId and InvoiceIssuerName between InvoiceDetail and CostAndUsage is defined on InvoiceDetail.BilledCost, as InvoiceDetail is the primary owner of invoice-level validation. CostAndUsage MAY reference it in its introductory description but MUST NOT restate it as a normative requirement.*

## 2. Dataset Requirements

> ### 2.1. Logical Grouping of Dataset Requirements

Author: Reminder:

Revisit/simplify this section in the guidelines and add a **Conformance to Attribute** section with alphabetical ordering.

<img width="1162" height="356" alt="image" src="https://github.com/user-attachments/assets/b2f1583e-b16f-46c4-a03e-df9d8c986364" />

| Technical        | Column Presence in Dataset                             | {DatasetId} MUST include {ColumnId} | N/A                                                        |
| Technical        | Technical Attributes Conformance                       | Always or when applicable                    | {DatasetId} MUST conform to ColumnHandling requirements.   |
| Business         | Business/Contextual Attributes Conformance             | When applicable                              | {DatasetId} MUST conform to DiscountHandling requirements. |
| Business         | Other Business/Contextual Requirements (FOR FUTURE USE)| For future use                               | N/A                                                        |

### 2.2. Ordering of Dataset Requirements Within Groups

To further enhance readability, individual requirements within each group SHOULD be ordered as follows:

* **MUST** – an absolute requirement
* **MUST NOT** – a prohibition
* **SHOULD** – recommended but not mandatory
* **SHOULD NOT** – discouraged but not strictly prohibited
* **MAY** – optional
* **MAY NOT** – optional prohibition / permitted not to

> > ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](#editorialstyleguidelines).*

Reviewer 7: ```suggestion
> ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](editorial-guidelines.md).*
```

`#editorialstyleguidelines` is a dead anchor (no such heading in this document). The editorial guidelines live at `editorial-guidelines.md` in the same directory. Same fix needed on lines 357 and 843.

  * <DatasetId> MUST be present[ when <Condition>].
  ```

### 4. Consistent Wording and Patterns in Dataset Requirements

Use standardized phrasing and terminology, and apply common requirement patterns where applicable to ensure clarity and consistency across datasets and corresponding requirements.

#### 4.1. Dataset Requirement Patterns

##### 4.1.1. Technical Requirements: Dataset Presence

```markdown

> * <DatasetId> MUST be present[ when <Condition>].

Reviewer 6: Should this condition get wrapped in brackets? Disregard if so!

Author: the brackets are meant to indicate that the condition is optional.

Reviewer: @ijurica This is an old thread, and I agree this is correct, with no action here.  Resolving.

* <DatasetId> SHOULD include <ColumnId> when <Condition>.
```

##### 2.4.1.3. Technical Requirements: Technical Attributes Conformance

```markdown
* <DatasetId> MUST conform to <TechnicalAttributeId> requirements.
```

##### 2.4.1.4. Business Requirements: Business/Contextual Attributes Conformance

```markdown
* <DatasetId> MUST conform to <BusinessAttributeId> requirements.
```

### 2.5. Dataset Normative Requirements Examples

#### 2.5.1. **Contract Commitment**

> ContractCommitment adheres to the following requirements:

Reviewer 7: ```suggestion
ContractCommitment MUST adhere to the following requirements:
```

Section 1.2 (line 93) defines the structural anchor as `<Entity> MUST adhere to the following requirements:`. All five dataset/column examples (lines 287, 295, 683, 699, 710) omit the MUST. The attribute examples in section 4.7 use the correct form.

Five one-line fixes, same pattern on each.

* ContractCommitment MUST be present when the provider supports *contract commitments*.
* ContractCommitment MUST conform to [ColumnHandling](#columnhandling) requirements.
* ContractCommitment MUST conform to [NullHandling](#nullhandling) requirements.

#### 2.5.2. **Cost and Usage**

> CostAndUsage adheres to the following requirements:

Reviewer 7: ```suggestion
CostAndUsage MUST adhere to the following requirements:
```

Same structural anchor fix. See comment on line 287.

## Column Requirements

### 1. Logical Grouping of Column Requirements

> Grouping and ordering of requirements ensure clarity, logical flow, and consistency across all columns, making related requirements easy to identify and follow. This structure should be maintained for consistency across the specification.

Reviewer 6: ```suggestion
Grouping and ordering of requirements ensures clarity, logical flow, and consistency across all columns, making related requirements easy to identify and follow. This structure should is maintained for consistency across the specification.
```

Author: - I was aiming for “Grouping and ordering” as plural, hence **ensure**.
- I used “should be maintained” to acknowledge that these groupings might not apply universally.

Author: Assuming the explanation above is sufficient, I’m resolving this thread. Feel free to reopen if you disagree.

* CostAndUsage MUST conform to [InvoiceHandling](#invoicehandling) requirements.
* CostAndUsage MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#datagenerator-calculatedsplitcostallocationhandling) requirements.

> ## 3. Column Requirements

Reviewer 3: Anchor contains a hyphen, but the convention for other attributes (`#columnhandling`, `#nullhandling`, `#discounthandling`) is lowercase with no separators.

```suggestion
* CostAndUsage MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#datageneratorcalculatedsplitcostallocationhandling) requirements.
```

Author: **_Note:_**

After the last merge from the `working_draft` branch, it seems that some of the previous comments and suggestions are no longer on the correct line. 😞 

I’m pretty sure this suggestion was referring to the following line, which has since been removed from the guideline:

``` markdown
* CostAndUsage MUST conform to [DataGeneratorCalculatedSplitCostAllocationHandling](#datagenerator-calculatedsplitcostallocationhandling) requirements.
```

---

**Feedback:**

Great catch! 

For creating anchor links in content under the `/specification` directory, we use [`TableOfContents.py`](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/guidelines/contributors/markdownpp-guidelines.md#1-enhanced-table-of-contents-tableofcontentspy), which removes the `-` character from headers. This particular issue is actually still present in the file, but the snippet shown here is just a copy/snippet as an example of similar cases across the specification.  

I suggest we address all those broken links that include `(#datagenerator-calculatedsplitcostallocationhandling)` together as part of the consistency review - see [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1878#issuecomment-4124412190) in Issue #1878.

@thecloudman, do you agree with this approach, and can we close this thread?

Reviewer: We have logged this as AI #2127.  @thecloudman can you confirm we can fix this in consistency review and close?

Reviewer: Resolving in deference to AI #2127 during consistency review.

| Business             | Cost Calculation and Relationships | When applicable                       | When {Condition}, {ColumnId} adheres to the following additional requirements:<br>  *{ColumnId} of a charge calculated based on other charges (e.g., when the ChargeCategory is "Tax") MUST be calculated based on the ContractedCost of those related charges.<br>* {ColumnId} of a charge unrelated to other charges (e.g., when the ChargeCategory is "Credit") MUST match the BilledCost. |
| Business             | Other                              | When applicable                       |                                                                                           |

### 3.2. Ordering of Column Requirements Within Groups

To further enhance readability, individual requirements within each group SHOULD be ordered as follows:

* **MUST** – an absolute requirement
* **MUST NOT** – a prohibition
* **SHOULD** – recommended but not mandatory
* **SHOULD NOT** – discouraged but not strictly prohibited
* **MAY** – optional
* **MAY NOT** – optional prohibition / permitted not to

> > ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](#editorialstyleguidelines).*

Reviewer 7: ```suggestion
> ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](editorial-guidelines.md).*
```

Same dead-link fix as line 234.

  * Column normative requirements specify requirements of the column such as nullability.
  * JSON schema normative requirements specify the shape of the JSON.
  * Contents normative requirements usually specify the expected Keys, the format of the Values, and the expected contents of the Values.

#### 4.2. JSON Schema

* **Omit JSON schema normative requirements for Key-Value Format columns**: The Key-Value Format definition is sufficient to define the expected JSON schema.

* **Include JSON schema normative requirements for JSON Object Format columns**: The JSON Object Format specifies that the format is subject to the requirement of the column and that data generator-defined columns must have documented schema.
  * The pattern used in [AllocatedMethodDetails](#allocatedmethoddetails) and [ContractApplied](#contractapplied) consists of Object containing a collection whose key is "Elements" which contains one or more objects in the Key-Value format.

>   **Example JSON**

Reviewer 5: 🤖 [AI][Claude] Minor: Consider adding a blank line before this header for consistency with markdown lint rules (MD031).

```suggestion

  **Example JSON**
```

Author: I think you’re probably referring to lines 393–394

Author: addressed 393-394. resolving this thread

* **Default to Singular Form**: Column references should be singular, with the understanding that the requirement applies to all values in the column.

### 3.4. Additional Guidelines for Columns in JSON Format

#### 3.4.1. Column Definition Structure

> * **Separate normative requirements into sections for column, JSON schema, and contents**: Communicating the normative requirements for a column, JSON schema, and the contents can be convoluted. Separating these requirements provides better clarity.

Reviewer: @ijurica 
We are not currently breaking out into these three areas for AllocatedMethodDetails, ContractApplied, or ContractCommitmentApplicability.  It's just column and object.

How should we revise this section to reflect that?

Reviewer: Tracking in consistency review.  Resolving.

      * <ColumnId> MUST be <SpecificRequirement>.
  ```

### 3.7. Consistent Wording and Patterns in Column Requirements

To ensure clarity and consistency across columns and corresponding requirements, it is important to:

* Follow common requirement patterns where applicable
* Use standardized phrasing and terminology

#### 3.7.1. Column Requirement Patterns

##### 3.7.1.1. Technical Requirements: Data Type

```markdown
* <ColumnId> MUST be of type String.
* <ColumnId> MUST be of type Decimal.
* <ColumnId> MUST be of type Date/Time.
```

##### 3.7.1.2. Technical Requirements: Value Format

```markdown
* <ColumnId> MUST conform to [StringHandling](#stringhandling) requirements.
* <ColumnId> MUST conform to [Numeric Format](#numericformat) requirements.
* <ColumnId> MUST conform to [DateTimeFormat](#date/timeformat) requirements.
* <ColumnId> SHOULD conform to [UnitFormat](#unitformat) requirements.
* <ColumnId> MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* <ColumnId> MUST conform to [CurrencyFormat](#currencyformat) requirements.

> ```

Reviewer 3: Three anchor issues: display name has a space (`Numeric Format` → `NumericFormat`), slash in anchor (`#date/timeformat` → `#datetimeformat`), and hyphen in anchor (`#key-valueformat` → `#keyvalueformat`). All should match the PascalCase-to-lowercase convention used by adjacent anchors.

```suggestion
* <ColumnId> MUST conform to [NumericFormat](#numericformat) requirements.
* <ColumnId> MUST conform to [DateTimeFormat](#datetimeformat) requirements.
* <ColumnId> SHOULD conform to [UnitFormat](#unitformat) requirements.
* <ColumnId> MUST conform to [KeyValueFormat](#keyvalueformat) requirements.
```

Author: You're absolutely right about the PascalCase-to-lowercase convention and the issues with spaces, and hyphens. Some of these inconsistencies still exist in the specification and should be addressed as part of the consistency review (recorded in [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/1878#issuecomment-4124412190) in #1878).

That said, in the latest revision of the guidelines I moved away from listing specific format attributes and instead introduced a more generic pattern:

``` markdown
* <ColumnId> MUST conform to <FormatAttributeId> requirements.
```

The intent here is to **reduce tight coupling between the guidelines and the specification**. By avoiding explicit references to individual format attributes (e.g., NumericFormat, DateTimeFormat, etc.), the guidelines become more maintainable and resilient to future changes in the spec.

---

@thecloudman, assuming this approach works for you, I’ll go ahead and resolve this thread. 
If you feel we should still include concrete examples, feel free to reopen.

> * <ColumnId> MUST conform to [CurrencyCodeFormat](#currencycodeformat) requirements.

Reviewer 7: ```suggestion
* <ColumnId> MUST conform to [CurrencyFormat](#currencyformat) requirements.
```

The spec heading is "Currency Format" (`currency_format.md`). There is no "CurrencyCodeFormat" attribute in the spec.

>   * <ColumnId> MUST be null in all other cases.

Reviewer 6: ```suggestion
* <ColumnId> MUST be null in all other cases.
```

Author: This was intentional, not an oversight. The pattern outlined here is applied exactly as shown, with the nested bullet. Closing this thread, but feel free to reopen if clarification is needed.

> ##### Business & Contextual Requirements: Cost Validation Rules

Reviewer: ### Markdown Header Mismatch

To keep the document consistent, this heading needs to be updated to match the newly defined requirement group name from the tabular overview in the "Logical Grouping of Column Requirements" section (which was changed from "Cost Validation Rules" to simply "Cost Validation"). 

**Suggested Fix:**
```suggestion
+##### Business & Contextual Requirements: Cost Validation
```

  * When the service provider has only one user-defined tag scheme. (instead of: When the service provider has only 1 user-defined tag scheme.)
  * When the service provider has more than one user-defined tag scheme. (instead of: When the service provider has 2 or more user-defined tag schemes.)

### 3.8. Column Normative Requirements Examples

#### 3.8.1. **List Unit Price**

> ListUnitPrice adheres to the following requirements:

Reviewer 7: ```suggestion
ListUnitPrice MUST adhere to the following requirements:
```

Same structural anchor fix. See comment on line 287.

* The sum of <ColumnId> in a given billing period MUST/MAY NOT match the sum of the invoices received for that billing period for a billing account.
```

##### Business & Contextual Requirements: Cost Validation

```markdown
* <CostColumnId> MUST equal the product of <UnitPriceColumnId> and PricingQuantity when <UnitPriceColumnId> is not null and PricingQuantity is not null.
```

#### Column Requirement Standardized Terminology

##### Identifiers and Uniqueness within Scope

* Patterns:

  ```markdown
  * {ColumnId} MUST be a unique identifier within {Scope}.
  * {ColumnId} SHOULD be a fully-qualified identifier.
  ```

* Examples:

  ```markdown
  * BillingAccountId MUST be a unique identifier within a service provider.
  * ResourceId SHOULD be a fully-qualified identifier.

>   ```

Reviewer: Several of the markdown code blocks in this section are indented. 

According to `AGENTS.md` under the File Organization rules: "Code blocks must be aligned to start of line (not indented)".

Removal of indentation needs to be applied through this section.

````suggestion
```markdown
* BillingAccountId MUST be a unique identifier within a service provider.
* ResourceId SHOULD be a fully-qualified identifier.
```
````

Author: **Question:** Does this rule apply to `guidelines/` files as well? In `AGENTS.md`, it's defined under `Writing Specification Content > File Organization`, which can be interpreted as applicable to `specification/` content that goes through the build pipeline.

For context, the code blocks here are intentionally indented to visually associate them with their parent bullet points.

Reviewer: You are correct that the `guidelines/` folder is currently excluded from the build pipeline and `pymarkdownlnt` scans. However, the point stands for the sake of **pattern modeling** and repository health:

* **Canonical Examples**: These guidelines provide the canonical examples that contributors (and AI agents) copy-paste into the specification. If we leave them indented here, they will inevitably be copy-pasted into the `specification/` folder, where they **will** trigger build failures due to the strict linting rules.
* **AI Agent Guidance**: `AGENTS.md` is specifically designed to guide AI coding assistants. AI models are highly sensitive to patterns in documentation; if the guidelines use indented code blocks, an AI may "learn" that indentation is the project standard, leading to frequent formatting errors in actual specification files.
* **Core Principles**: Modeling the "flush-left" requirement here ensures that the patterns we ask people to follow are technically safe and consistent across the entire repository, supporting the design note that "Consistency helps with clarity".

In my opinion, aligning all code blocks to the start of the line ensures our documentation models the exact state we require in the final specification.

Author: addressed. 

@shawnalpay, can you please check when you get a chance. thanks!

Reviewer: Looks good, thanks!

  * The product of ListUnitPrice and [PricingQuantity](#pricingquantity) MUST match the [ListCost](#listcost) when PricingQuantity is not null and ChargeClass is not "Correction".
  * Discrepancies in ListUnitPrice, ListCost, or PricingQuantity MAY exist when ChargeClass is "Correction".

#### 3.8.2. **Billed Cost**

> BilledCost adheres to the following requirements:

Reviewer 7: ```suggestion
BilledCost MUST adhere to the following requirements:
```

Same structural anchor fix. See comment on line 287.

* BilledCost MUST be denominated in the BillingCurrency.

> * The sum of BilledCost in a given [*billing period*](#glossary:billing-period) MUST match the sum of the invoices received for that *billing period* for a [*billing account*](#glossary:billing-account).

Reviewer 7: ```suggestion
* The sum of BilledCost for a given [InvoiceId](#invoiceid) MUST match the sum of the payable amount provided in the corresponding invoice with the same id generated by the [InvoiceIssuerName](#invoiceissuername).
```

The current wording uses billing period + billing account as the grouping key. The working draft spec (`billedcost.md`) now uses InvoiceId. This is a material difference in the reconciliation logic.

Author: I'll commit this suggestion for now, but in the end I’ll likely choose a different column instead of `BilledCost`, i.e., one that won’t have changes in this release. 😄

#### 3.8.3. **CommitmentDiscountQuantity**

> CommitmentDiscountQuantity adheres to the following requirements:

Reviewer 7: ```suggestion
CommitmentDiscountQuantity MUST adhere to the following requirements:
```

Same structural anchor fix. See comment on line 287.

### 2. Structural Anchor Requirement for Attributes

Each Attribute Requirements section MUST begin with a structural anchor requirement.

The structural anchor requirement:

* uses the Attribute ID as the subject,
* introduces the scope of the requirements,
* is non-verifiable and non-enforceable,
* exists solely for structural consistency and automated parsing.

Canonical form:

``` markdown
<AttributeId> MUST adhere to the following requirements:
```

This is the only case in which an Attribute ID may appear as a normative subject.

### 3. Normative Subjects in Attribute Requirements

All enforceable normative requirements within an Attribute section MUST target schema-level entities that conform to the Attribute. The Attribute itself MUST NOT be treated as an enforceable subject beyond the structural anchor requirement.

Each Attribute implicitly assumes one or more intended normative subjects. These subjects determine which schema-level entities the Attribute’s requirements apply to.

The following table lists commonly used intended subjects.

***Note:** This list is non-exhaustive and will be extended over time.*

> | Subject (Core) | Distinct Subjects with Qualifiers | Description/Note |

Reviewer 5: 🤖 [AI][Claude] The table includes "Attribute" with `<Attribute ID>` as a subject (first data row after header), but Section 3 states "The Attribute itself MUST NOT be treated as an enforceable subject beyond the structural anchor requirement." Consider adding a note in the Description column clarifying this refers only to structural anchors, or moving this row to a separate "Special Cases" section to avoid confusion.

Author: Check out the `### 2. Structural Anchor Requirement` section`:

> For **Attribute Requirements** sections, the Attribute ID MAY be used as the subject of the structural anchor requirement.
> This usage is a structural exception only and MUST NOT be interpreted as implying that Attributes are normative subjects or independently enforceable schema entities.

`#### 3.1 Allowed Subjects` section as well
`
> **Exception:** A structural anchor requirement MAY use the ID of the enclosing schema construct as its subject, even if that construct is not otherwise an allowed normative subject (e.g., `<FOCUS Attribute ID>`). This exception applies only to structural anchor requirements.

Etc.

Author: this content has been removed from the document so I'm resolving this thread.

| Key value      | Key values in Object in FOCUS/Custom column containing JsonObjectFormat values | |
| Key            | Keys in FOCUS/Custom column containing values in key-value pair format | |
| Key value      | Key values in FOCUS/Custom column containing values in key-value pair format | |

### 4. Grouping of Attribute Requirements

Grouping and ordering of Attribute requirements ensure clarity, consistency, and maintainability across the specification, making related requirements easy to identify and follow.

Attributes may include requirements that apply to one or more intended normative subjects. To make the applicability of each Attribute, and each of its individual requirements, as transparent as possible, intended normative subjects serve as the basis for grouping. This ensures that readers can readily determine whether a requirement applies to a dataset, a subset of datasets, FOCUS columns, or custom columns.

This structured grouping improves clarity, consistency, and maintainability across the specification by making related requirements easier to locate and understand, without introducing any additional normative meaning.

> 0. **Structural Attribute Anchor Requirement:** Introduces the scope of the Attribute and provides a stable parsing entry point; it does not introduce a verifiable constraint.

Reviewer 5: 🤖 [AI][Claude] Section numbering starts at "0." - is this intentional to indicate the structural anchor is a special preamble? If so, this is a nice touch! If not, consider renumbering to start at "1." for consistency.

Author: intentional 😄

Reviewer 6: +1 this. The numbering doesn't feel user friendly to me. However, it's a preference vs. something that I feel needs to be changed. The section to follow, I feel would benefit from following minor release format, or whatever its numerical system name is.

Author: I here you. 

Markdown numbered lists tend to get a bit tricky once nesting is involved, which is partly why the numbering currently looks like this. 😞 
I’ll revisit and refine the numbering soon. Considering switching to a nested structure consistently across the document.

Resolving this thread.

The structural anchor requirement:

* uses the Attribute ID as the subject,
* introduces the scope of the requirements,
* is non-verifiable and non-enforceable,
* exists solely for structural consistency and automated parsing.

Canonical form:

``` markdown
<AttributeId> MUST adhere to the following requirements:
```

This is the only case in which an Attribute ID may appear as a normative subject.

### 3. Normative Subjects in Attribute Requirements

All enforceable normative requirements within an Attribute section MUST target schema-level entities that conform to the Attribute. The Attribute itself MUST NOT be treated as an enforceable subject beyond the structural anchor requirement.

Each Attribute implicitly assumes one or more intended normative subjects. These subjects determine which schema-level entities the Attribute’s requirements apply to.

The following table lists commonly used intended subjects.

***Note:** This list is non-exhaustive and will be extended over time.*

| Subject (Core) | Distinct Subjects with Qualifiers | Description/Note |
|----------------|-----------------------------------|------------------|

> | Attribute      | `<Attribute ID>` | Specific Attribute identified by Dataset ID |

Reviewer 7: ```suggestion
| Attribute      | `<Attribute ID>` | Specific Attribute identified by Attribute ID |
```

Description says "identified by Dataset ID" but this row is about Attributes. Copy-paste error.

Author: After we agreed that requirements targeting specific datasets and columns should not be kept at the Attribute level, this content (this line) has been removed from the document.

resolving this thread

| Column         | FOCUS/Custom column representing national currency | |
| Column         | FOCUS/Custom column representing virtual currency | |
| Column         | FOCUS/Custom column representing an identifier | |
| Column         | FOCUS/Custom column representing a name | |
| Column         | FOCUS/Custom column representing a product offering that incurred the charge | |
| Column         | FOCUS column with `Category` suffix | |
| Column         | FOCUS/Custom column containing numeric values | |
| Column         | FOCUS/Custom column containing date/time values | |
| Column         | FOCUS/Custom column containing string values | |
| Column         | FOCUS/Custom column containing immutable string values | |
| Column         | FOCUS/Custom column containing mutable string values | |
| Column         | FOCUS/Custom column containing not-nullable string values | |
| Column         | FOCUS/Custom column containing values in key-value pair format | |
| Key            | Keys in FOCUS/Custom column containing values in key-value pair format | |
| Key value      | Key values in FOCUS/Custom column containing values in key-value pair format | |

> | Column         | FOCUS column containing values in JsonObjectFormat format | |

Reviewer 7: ```suggestion
| Column         | FOCUS column containing JsonObjectFormat values | |
```

"JsonObjectFormat format" reads as "format format." Lines 786-789 use "JsonObjectFormat values" without the extra "format."

Author: Since we no longer use subject qualifiers in this form and do not include them in this document, I'm resolving this thread

Note: I like the suggestion and I’m noting it down on the side to make sure I don’t accidentally use the old wording in PR #1981. 🙂

* Requirements targeting the general subject (e.g., FOCUS column) first
* Requirements targeting qualified subjects (e.g., FOCUS column containing numeric values) after
* Within each subject, order by BCP 14 keyword:
  * **MUST** – an absolute requirement
  * **MUST NOT** – a prohibition
  * **SHOULD** – recommended but not mandatory
  * **SHOULD NOT** – discouraged but not strictly prohibited
  * **MAY** – optional
  * **MAY NOT** – optional prohibition / permitted not to

> ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](editorial-guidelines.md).*

* For detailed interpretation of keywords such as "MUST", "MUST NOT", "SHOULD", "SHOULD NOT", "MAY", and others, see [BCP14](https://tools.ietf.org/html/bcp14) [[RFC2119](https://tools.ietf.org/html/rfc2119)][[RFC8174](https://tools.ietf.org/html/rfc8174)].

### 4.7. Attribute Normative Requirements Examples

> #### 4.7.1. Null Handling

Reviewer 7: ```suggestion
**Notes:**

* The examples below are **snippets** that illustrate guideline-compliant patterns, not full listings. Some examples reformulate existing requirements; others show aspirational patterns not yet in the specification.
* Authors should consult the actual FOCUS attribute specification files as the **source of truth**, as these guidelines may not always reflect the latest version.

#### 4.7.1. Null Handling
```

Sections 2.5 (line 288) and 3.8 (line 688) both include this disclaimer. Section 4.7 is the only example section without one, and it's the section most likely to be misread as verbatim spec since the attribute examples are deliberately restructured.

Author: Suggestion accepted; Unable to commit directly because the comments were no longer aligned with the target lines after a merge. Resolving

>   * When ChargeCategory is "Usage":

Reviewer 3: The spec file `commitmentdiscountquantity.md` uses *term* for commitment discount duration, not *period*. The glossary link `[*period*](#glossary:period)` references a different concept.

```suggestion
    * CommitmentDiscountQuantity MUST be the quantity of CommitmentDiscountUnit, paid fully or partially upfront, that is eligible for consumption over the *commitment discount's* *term* when [ChargeFrequency](#chargefrequency) is "One-Time".
```

Author: this suggestion was applied in [this commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/commits/36bc1ddfb76812d31ea1e6398db7368b8399e85f).
Resolving this thread

The structural anchor requirement:

* uses the primary schema-level entity as the subject,
* references the Attribute ID to establish the conformance context,
* introduces the scope of the subsequent requirements,
* is non-verifiable and non-enforceable,
* exists solely for structural consistency and automated parsing.

The canonical form of the structural anchor requirement is:

```markdown
[Dataset|Column] conforming to <Attribute ID> attribute MUST adhere to the following requirements:
```

Where `[Dataset|Column]` is the primary schema-level entity targeted by the Attribute — either Dataset or Column. Most Attributes target either datasets or columns, but not both. When an Attribute targets both datasets and columns, a separate structural anchor requirement MUST be used for each entity type.

When an Attribute is applicable only under specific conditions, the structural anchor requirement MAY be preceded by an applicability criteria condition:

```markdown
When <actor> <applicability-citeria-condition>, [Dataset|Column] conforming to <Attribute ID> attribute MUST adhere to the following requirements:
```

> #### 4.3. Normative Subjects in Attribute Requirements

Reviewer 7: ```suggestion
### 4.3. Normative Subjects in Attribute Requirements
```

Heading level is `####` but peer sections (4.1, 4.2, 4.5, 4.6, 4.7) are all `###`.

When an Attribute's requirements do not apply to all entities within scope but only to a subset, a qualifier condition narrows the scope by describing that subset (e.g., `When FOCUS column contains numeric values, FOCUS column MUST adhere to the following requirements`). This ensures that the applicability of each requirement is explicit and does not rely solely on the conformance declaration.

The following table provides an overview of anchor subject types and requirement subjects used across all attributes, with each Attribute typically targeting only a subset of these subjects.

| Anchor Subject Type | Requirement Subject |
|---|---|
| Dataset | FOCUS dataset |
| Column | FOCUS column |
| Column | Custom column |
| Column | Object in FOCUS column |
| Column | Object in array in FOCUS column |
| Column | Key in Object in FOCUS column |
| Column | Key value in Object in FOCUS column |
| Column | Key in FOCUS column |
| Column | Key value in FOCUS column |

> #### 4.4. FOCUS vs Custom Column Requirements

Reviewer 7: ```suggestion
### 4.4. FOCUS vs Custom Column Requirements
```

Same heading level fix as 4.3 above.

Custom columns are expected to behave in the same or similar way as FOCUS columns. The difference is that requirements for Custom columns are expressed as recommendations rather than mandatory constraints.

As a general rule, when authoring requirements for a FOCUS column, the author SHOULD also consider whether the same requirements apply to Custom columns and, where applicable, restate them as recommendations using `SHOULD` instead of `MUST` and `SHOULD NOT` instead of `MUST NOT`. `MAY` and `MAY NOT` SHOULD be retained as-is.

The following exceptions apply:

* Some requirements MAY not be applicable to Custom columns and SHOULD be omitted.
* Some requirements MAY be mandatory for Custom columns and SHOULD retain `MUST` or `MUST NOT`.
* Some requirements MAY be optional for Custom columns and SHOULD use `MAY` or `MAY NOT` instead.

> The appropriate keyword SHOULD always be chosen based on the intent and context of the requirement.

Reviewer 3: `MAY not` reads as informal English here ("might not"), but `MAY` is a BCP 14 keyword — `MAY NOT` would mean a normative optional prohibition, which isn't the intent. Using lowercase `might` avoids the ambiguity.

```suggestion
* Some requirements might not be applicable to Custom columns and SHOULD be omitted.
* Some requirements might be mandatory for Custom columns and SHOULD retain `MUST` or `MUST NOT`.
* Some requirements might be optional for Custom columns and SHOULD use `MAY` or `MAY NOT` instead.
```

Author: After the last merge from the `working_draft` branch, it seems that some of the previous comments and suggestions are no longer on the correct line. :(

Anyway, this suggestion was previously applied - see [this commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1967/commits/36bc1ddfb76812d31ea1e6398db7368b8399e85f)

Attaching current version:

> The following exceptions apply:
> 
> * Some requirements might not be applicable to Custom columns and SHOULD be omitted.
> * Some requirements might be mandatory for Custom columns and SHOULD retain `MUST` or `MUST NOT`.
> * Some requirements might be optional for Custom columns and SHOULD use `MAY` or `MAY NOT` instead.
> 
> The appropriate keyword SHOULD always be chosen based on the intent and context of the requirement.

And resolving the thread

   3. **Key values in Objects in Columns containing JsonObjectFormat values**
   4. **Keys in Columns containing Key-Value pair format values**
   5. **Key values in Columns containing Key-Value pair format values**
4. **Custom Column Attribute Requirements:**
   1. **Global Custom Column Requirements:** Applicable to all Custom columns, regardless of their structure or purpose.
   2. **Qualified Custom Column Requirements:** Applicable to a subset of Custom columns, identified through a qualifier.

### 4.6. Ordering of Attribute Requirements within Groups

To further enhance readability, individual requirements within each group SHOULD be ordered as follows:

* Requirements targeting the general subject (e.g., FOCUS column) first
* Requirements targeting qualified subjects (e.g., FOCUS column containing numeric values) after
* Within each subject, order by BCP 14 keyword:
  * **MUST** – an absolute requirement
  * **MUST NOT** – a prohibition
  * **SHOULD** – recommended but not mandatory
  * **SHOULD NOT** – discouraged but not strictly prohibited
  * **MAY** – optional
  * **MAY NOT** – optional prohibition / permitted not to

> > ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](#editorialstyleguidelines).*

Reviewer 7: ```suggestion
> ***Important Note:*** *The term **RECOMMENDED** (recommended but not mandatory; previously used only for presence-related normative requirements) is no longer permitted for use in normative requirements as of December 2025. The keyword **SHOULD** must be used instead. Please refer to the [**Editorial Style Guidelines**](editorial-guidelines.md).*
```

Same dead-link fix as line 234.

* **Column level:** A specific column declares conformance directly (e.g., `BilledCost MUST conform to NumericFormat requirements.`).

Normative requirements defined in an Attribute section are evaluated within the scope of conforming entities but apply only to the subjects explicitly defined by each requirement. Conformance determines the set of entities in scope, while the requirement subject determines which of those entities are targeted.

### Structural Anchor Requirement for Attributes

Each Attribute Requirements section MUST begin with a structural anchor requirement.

The structural anchor requirement:

* uses the primary schema-level entity as the subject,
* references the Attribute ID to establish the conformance context,
* introduces the scope of the subsequent requirements,
* is non-verifiable and non-enforceable,
* exists solely for structural consistency and automated parsing.

The canonical form of the structural anchor requirement is:

```markdown

> [Dataset|Column] conforming to <Attribute ID> attribute MUST adhere to the following requirements:

Reviewer: ```suggestion
[Dataset|Column] conforming to <AttributeId> attribute MUST adhere to the following requirements:
```

```

Where `[Dataset|Column]` is the primary schema-level entity targeted by the Attribute — either Dataset or Column. Most Attributes target either datasets or columns, but not both. When an Attribute targets both datasets and columns, a separate structural anchor requirement MUST be used for each entity type.

When an Attribute is applicable only under specific conditions, the structural anchor requirement MAY be preceded by an applicability criteria condition:

```markdown

> When <actor> <applicability-citeria-condition>, [Dataset|Column] conforming to <Attribute ID> attribute MUST adhere to the following requirements:

Reviewer: ```suggestion
When <actor> <applicability-criteria-condition>, [Dataset|Column] conforming to <AttributeId> attribute MUST adhere to the following requirements:
```
